File: src/main/java/ghx/MotionDetector.java
package ghx;

import java.util.ArrayList;
import java.util.List;

public class MotionDetector {
    private ImageSource imageSource;
    private final List<Listener> listeners = new ArrayList<>();

    public MotionDetector(ImageSource imageSource) {
        this.imageSource = imageSource;
    }

    public void start() {

    }

    void addListener(Listener l) {
        listeners.add(l);
    }

    interface Listener {
        void onMotionDetection();
    }
}

File: src/main/scala/ghx/MotionDetector.scala
package ghx

import java.util

import scala.collection.mutable

class MotionDetector {
  private var imageSource: ImageSource = null
  private val listeners: mutable.ListBuffer[MotionDetector.Listener] = new mutable.ListBuffer[MotionDetector.Listener].empty

  def this(imageSource: ImageSource) = {
    this()
    this.imageSource = imageSource
  }

  def start(): Unit = {
  }

  private[ghx] def addListener(l: MotionDetector.Listener): Unit = {
    listeners.addOne(l)
  }

}

object MotionDetector {
  trait Listener {
    def onMotionDetection(): Unit
  }
}
